Source/Core/VideoCommon/VertexLoaderManager.cpp

```diff
@@ -198,57 +197,45 @@ NativeVertexFormat* GetUberVertexFormat(const PortableVertexDeclaration& decl)
-  g_current_vat = vtx_attr_group;
+  CPState* state = IsPreprocess ? &g_preprocess_cp_state : &g_main_cp_state;
+  BitSet8& attr_dirty = IsPreprocess ? g_preprocess_vat_dirty : g_main_vat_dirty;
+  auto& vertex_loaders = IsPreprocess ? g_preprocess_vertex_loaders : g_main_vertex_loaders;
 
   VertexLoaderBase* loader;
-  if (attr_dirty[vtx_attr_group])
+
+  // We are not allowed to create a native vertex format on preprocessing as this is on the wrong
+  // thread
+  bool check_for_native_format = !IsPreprocess;
+
+  VertexLoaderUID uid(state->vtx_desc, state->vtx_attr[vtx_attr_group]);
+  std::lock_guard<std::mutex> lk(s_vertex_loader_map_lock);
+  VertexLoaderMap::iterator iter = s_vertex_loader_map.find(uid);
+  if (iter != s_vertex_loader_map.end())
   {
-    // We are not allowed to create a native vertex format on preprocessing as this is on the wrong
-    // thread
-    bool check_for_native_format = !preprocess;
-
-    VertexLoaderUID uid(state->vtx_desc, state->vtx_attr[vtx_attr_group]);
-    std::lock_guard<std::mutex> lk(s_vertex_loader_map_lock);
-    VertexLoaderMap::iterator iter = s_vertex_loader_map.find(uid);
-    if (iter != s_vertex_loader_map.end())
-    {
-      loader = iter->second.get();
-      check_for_native_format &= !loader->m_native_vertex_format;
-    }
-    else
-    {
-      s_vertex_loader_map[uid] =
-          VertexLoaderBase::CreateVertexLoader(state->vtx_desc, state->vtx_attr[vtx_attr_group]);
-      loader = s_vertex_loader_map[uid].get();
-      INCSTAT(g_stats.num_vertex_loaders);
-    }
-    if (check_for_native_format)
-    {
-      // search for a cached native vertex format
-      const PortableVertexDeclaration& format = loader->m_native_vtx_decl;
-      std::unique_ptr<NativeVertexFormat>& native = s_native_vertex_map[format];
-      if (!native)
-        native = g_renderer->CreateNativeVertexFormat(format);
-      loader->m_native_vertex_format = native.get();
-    }
-    vertex_loaders[vtx_attr_group] = loader;
-    attr_dirty[vtx_attr_group] = false;
+    loader = iter->second.get();
+    check_for_native_format &= !loader->m_native_vertex_format;
   }
   else
   {
-    loader = vertex_loaders[vtx_attr_group];
+    s_vertex_loader_map[uid] =
+        VertexLoaderBase::CreateVertexLoader(state->vtx_desc, state->vtx_attr[vtx_attr_group]);
+    loader = s_vertex_loader_map[uid].get();
+    INCSTAT(g_stats.num_vertex_loaders);
   }
-
-  // Lookup pointers for any vertex arrays.
-  if (!preprocess)
-    UpdateVertexArrayPointers();
-
-  return loader;
+  if (check_for_native_format)
+  {
+    // search for a cached native vertex format
+    const PortableVertexDeclaration& format = loader->m_native_vtx_decl;
+    std::unique_ptr<NativeVertexFormat>& native = s_native_vertex_map[format];
+    if (!native)
+      native = g_renderer->CreateNativeVertexFormat(format);
+    loader->m_native_vertex_format = native.get();
```

Comment:
Couldn't this entire block just be:

`loader->m_native_vertex_format = GetOrCreateMatchingFormat(loader->m_native_vtx_decl)`

?

Reply:
FYI I just moved this code around, I didn't write it.

Reply:
Shall I clean it up anyway?

Reply:
Yeah sorry, I noticed that after my comment.  I think it'd be worthwhile, not sure if any of the other devs have concerns about it muddying the changes.

Reply:
Feel free to clean it up; the existing code is definitely a mess. (I'd recommend against changing the vertex loader JITs themselves, though, as I'm doing a bit of that in #10808.)

Reply:
@iwubcode The current code doesn't store the newly created format in the map. I don't know if there's a good reason to not do that.